tests/core/arch-x86/stalker-x86.c

```diff
@@ -375,6 +401,239 @@ invoke_flat (TestStalkerFixture * fixture,
+  g_assert_cmpuint (runner_range.size, !=, 0);
+
+  /* Initialize stalker */
+  gum_stalker_set_trust_threshold (fixture->stalker, 3);
+  trust = gum_stalker_get_trust_threshold (fixture->stalker);
+
+  /* Create IPC */
+  ret = pipe2 (compile_pipes, O_NONBLOCK);
+  g_assert_cmpint (ret, >=, 0);
+
+  ret = pipe2 (execute_pipes, O_NONBLOCK);
+  g_assert_cmpint (ret, >=, 0);
+
+  // fcntl (execute_pipes[STDIN_FILENO], F_SETPIPE_SZ, 1UL << 20);
+
+  /* Configure stalker */
+  g_clear_object (&fixture->transformer);
+  fixture->transformer = gum_stalker_transformer_make_from_callback (
+      prefetch_transformer, NULL, NULL);
+
+  gum_stalker_follow_me (fixture->stalker, fixture->transformer,
+      GUM_EVENT_SINK (fixture->sink));
+  gum_stalker_deactivate (fixture->stalker);
+
+  /* Run the child */
+  prefetch_run_child (fixture->stalker, &runner_range,
+      compile_pipes[STDOUT_FILENO], execute_pipes[STDOUT_FILENO]);
+
+  /* Read the results */
+  compiled_run1 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (compile_pipes[STDIN_FILENO], compiled_run1);
+  executed_run1 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (execute_pipes[STDIN_FILENO], executed_run1);
+
+  compiled_size_run1 = g_hash_table_size (compiled_run1);
+  g_print ("\tcompiled: %d\n", compiled_size_run1);
+
+  executed_size_run1 = g_hash_table_size (executed_run1);
+  g_print ("\texecuted: %d\n", executed_size_run1);
+
+  g_assert_cmpuint (compiled_size_run1, >, 0);
+  g_assert_cmpuint (compiled_size_run1, ==, executed_size_run1);
+
+  /* Prefetch the blocks */
+  gum_stalker_set_trust_threshold (fixture->stalker, 0);
+  g_hash_table_iter_init (&iter, compiled_run1);
+  while (g_hash_table_iter_next (&iter, &iter_key, &iter_value))
+  {
+    for (i = 0; i < trust; i++)
+      gum_stalker_prefetch (fixture->stalker, iter_key);
+  }
+  gum_stalker_set_trust_threshold (fixture->stalker, trust);
+
+  /* Run the child again */
+  prefetch_run_child (fixture->stalker, &runner_range,
+      compile_pipes[STDOUT_FILENO], execute_pipes[STDOUT_FILENO]);
+
+  /* Read the results */
+  compiled_run2 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (compile_pipes[STDIN_FILENO], compiled_run2);
+  executed_run2 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (execute_pipes[STDIN_FILENO], executed_run2);
+
+  compiled_size_run2 = g_hash_table_size (compiled_run2);
+  g_print ("\tcompiled2: %d\n", compiled_size_run2);
+
+  executed_size_run2 = g_hash_table_size (executed_run2);
+  g_print ("\texecuted2: %d\n", executed_size_run2);
+
+  g_assert_cmpuint (compiled_size_run2, ==, 0);
+  g_assert_cmpuint (executed_size_run2, ==, executed_size_run1);
+
+
+  /* Free resources */
+  g_hash_table_destroy (compiled_run2);
+  g_hash_table_destroy (executed_run2);
+  g_hash_table_destroy (compiled_run1);
+  g_hash_table_destroy (executed_run1);
+
+  gum_stalker_unfollow_me (fixture->stalker);
+
+  g_print ("DONE\n");
+
+}
```

Comment:
Are pipes leaked? (Any other resource leaks?)

Reply:
Well spotted. I think that's it? The hash tables are freed, and the transformer is set in the fixture which is tidied up when the fixture is torn down.